docs/source/DeveloperGuide/Python3Migration/index.rst

```diff
@@ -0,0 +1,227 @@
+  This exposes subtle issues in Python 2 that were likely never noticed and an automatic conversion to fix this is inherently impossible.
+  More details about this can be found `here <https://portingguide.readthedocs.io/en/latest/strings.html>`_.
+
+  In most situations DIRAC is only dealing with ascii or unicode strings and therefore nothing needs to change.
+  However many libraries choose to be independent of the character encoding used and therefore return a ``bytes`` object in Python 3 instead of ``str``
+
+  .. code-block:: python
+
+    result = subprocess.check_output(["echo", "Hello"])
+
+    # Bad: Fails on Python 3 with "TypeError: can't concat str to bytes"
+    return "Result is" + result
+
+    # Good: Explicitly decode bytes to str (does nothing on Python 2)
+    return "Result is" + result.decode()
+
+    # For subprocess functions, the universal_newlines=True argument can be used
+    other_result = subprocess.check_output(["echo", "Hello"], universal_newlines=True)
+    # Good: other_result is already a str object
+    return "Result is" + other_result
+
+  **Checking the type of a string:**
+
+  .. code-block:: python
+
+    # Bad: Types should be check using isinstance
+    if type(my_variable) == str:
+
+    # Bad: basestring does not exist in Python 3
+    if instance(my_variable, basestring):
+
+    # Good: Supports both Python 2 and 3
+    if instance(my_variable, six.string_types):
+
+  **Reading files:**
+
+  It's preferable to explicitly state if a file is being opened in text mode or binary mode.
+
+  .. code-block:: python
+
+    # Bad: Works but it is unclear if data is expected to bytes or a string
+    with open("my_file.txt") as fp:
+       data = fp.read().split("\n")
+
+    # Good: File is explicitly in text mode
+    with open("my_file.txt", "rt") as fp:
+       data = fp.read().split("\n")
+
+    # Bad: Fails on python 3 as "\n" is a string not bytes
+    with open("my_file.txt", "rb") as fp:
+       data = fp.read().split("\n")
+
+    # Good: Prefix the "\n" to make it a bytes object
+    with open("my_file.txt", "rb") as fp:
+       data = fp.read().split(b"\n")
+
+**Dictionaries**
+  In Python 3 ``my_dict.keys()``, ``my_dict.values()`` and ``my_dict.items()`` now return an iterator instead of a list.
+  This is equivalent to ``my_dict.iterkeys()``, ``my_dict.itervalues()`` and ``my_dict.iteritems()`` in Python 2 and these methods have been removed.
+
+  In almost all cases ``my_dict``, ``my_dict.values()`` and ``my_dict.items()`` should be preferred.
+  The is a small overhead in Python 2 when using ``items()`` instead of ``iteritems()`` however this is only applicable when dealing with large dictionaries in tight loops and such code can likely be written as a faster alternative (``six`` provides functions like ``six.iteritems(my_dict)`` if absolutely necessary).
+
+  In rare cases the list object returned might be desirable, if so ``list(my_dict.items())`` can be used.
```

Comment:
so basically you are saying we have to bare the cost of a double copy as long as we use python 2 ?

Reply:
Does the cost actually matter? If it does, then fair enough and we can litter the code with `six.iterxxxx` but I doubt the cost of making the code less readable is worth it in most places.

Reply:
I thought that the solution was to use iteritems() everywhere in Python2 and list(obj.iteritems()) if the list is needed. When moving to Python3 we will globally substitute iteritems() by items().

Reply:
There is no benefit to using `list(obj.iteritems())` instead of `six.iteritems(obj)`.

As I state in the text of this PR, that makes it impossible to have a code base that supports both so we can gradually test services with Python 3 and fix bugs as we find them. I've never read about a project successfully migrating in one shot, there are just too many possible bugs.

Reply:
I think our original idea was to make things explicit when we want a copy. So the only way to achieve it which would be py2 and py3 compatible would be `items` when we do not need a copy, and `list(dict.items())` when we need one.

Reply:
Sorry I've misread your original comment as "bare the cost of a copy" instead of "bare the cost of a **double** copy".

I've tried to benchmark it and of course there is a cost but it's less than I would have naively expected. If there are a few hot spots where we have particularly big objects we could use `six.iteritems(list(obj))`.

Python 3 `list(x.items())` is notably faster than Python 2 `x.items()` so if there is a service where it matters it's a reason to move that service to Python 3 (or even PyPy which is a LOT faster).